File: LeetCode/208-implement-trie-prefix-tree.py
###############################################################################################
# typical trie application
###########
# 时间复杂度：O(n*2000), n is number of insert, search, and startswith
# 空间复杂度：O(n*2000*26)
###############################################################################################
class Trie:

    def __init__(self):
        self.sons = [[0]*26 for _ in range(2)]
        self.cnt = [0]*2
        self.idx = 1

    def insert(self, word: str) -> None:
        p = 0
        for i in range(len(word)):
            id_ = ord(word[i]) - ord('a')
            if not self.sons[p][id_]:
                self.sons[p][id_] = self.idx
                self.idx += 1
                # dynamic increasing
                self.sons.append([0]*26)
                self.cnt.append(0)
            p = self.sons[p][id_]
        self.cnt[p] += 1

    def search(self, word: str) -> bool:
        p = 0
        for i in range(len(word)):
            id_ = ord(word[i]) - ord('a')
            if not self.sons[p][id_]:
                return False
            p = self.sons[p][id_]
        return self.cnt[p]

    def startsWith(self, prefix: str) -> bool:
        p = 0
        for i in range(len(prefix)):
            id_ = ord(prefix[i]) - ord('a')
            if not self.sons[p][id_]:
                return False
            p = self.sons[p][id_]
        return True

File: LeetCode/208-implement-trie-prefix-tree.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// good implementation with class instead of vector
////////////////
// Time Complexity: O(n*2000), n is num of calls
// Space Complexity: O(n*2000)
///////////////////////////////////////////////////////////////////////////////////////////////
namespace my{
    class TreeNode{
        private:
            bool end;
            vector<TreeNode*> child;
        public:
            TreeNode(){
                this->end = false;
                child.resize(26,NULL);
            }

            void setEnd(){
                this->end = true;
            }

            void setNode(char ch){
                child[ch-'a'] = new TreeNode();
            }

            bool isEnd(){
                return this->end;
            }

            bool isNode(char ch){
                return child[ch-'a'] != NULL;
            }

            TreeNode* getNode(char ch){
                return child[ch-'a'];
            }
    };
}

class Trie {
private:
    my::TreeNode* root;
public:
    Trie() {
        root = new my::TreeNode();
    }
    
    void insert(string word) {
        my::TreeNode* p = root;
        for(int i = 0; i < word.size(); i++){
            char ch = word[i];
            if(p->isNode(ch)){
                p = p->getNode(ch);
            }else{
                p->setNode(ch);
                p = p->getNode(ch);
            }
        }
        p->setEnd();
    }
    
    bool search(string word) {
        my::TreeNode* p = root;
        for(int i = 0; i < word.size(); i++){
            char ch = word[i];
            if(p->isNode(ch)){
                p = p->getNode(ch);
            }else{
                return false;
            }
        }
        return p->isEnd();
    }
    
    bool startsWith(string prefix) {
        my::TreeNode* p = root;
        for(int i = 0; i < prefix.size(); i++){
            char ch = prefix[i];
            if(p->isNode(ch)){
                p = p->getNode(ch);
            }else{
                return false;
            }
        }
        return true;
    }
};